### DISCUSSION TITLE
Call for ideas: the mechanism for recovering an unfinished installation

### DISCUSSION THREAD

**Discussion by pzhlkj6612** [miurahr/aqtinstall#422] [Ideas]
Prerequisites:

- #421

Source: https://github.com/miurahr/aqtinstall/pull/420#discussion_r721494482

**Comment by ddalcino** [comment]
I think that to enable this feature properly, you would need to modify `aqt` to print or log a lot more information about its state in the event of a crash. Specifically:

1. Which archives have been downloaded and extracted successfully
2. Which archives have been downloaded but not extracted
3. Which archives have not been downloaded yet
4. (Optional) Which archives have been partially downloaded, and could possibly be resumed

We would also need an independent mechanism for running the Updater after everything has been extracted to the proper location. I don't think it's reasonable to expect any user to be able to patch all these binary files manually. This tool would also need some mechanism that ensures that all the required archives have been downloaded and extracted successfully; you don't want it to crash if it can't patch a file that doesn't exist, and our users will blame us if it doesn't work.

I think that if implemented, this feature should write something like an `aqt-recovery.json` file that contains all of the above information when `aqt` fails. We could add a separate `aqt resume` command that reads this json file and automatically picks up where `aqt` left off the last time.

This feature would need to be tested exhaustively. You would need to understand all the ways that `aqt` can fail, and make sure that in every one, `aqt` writes the `aqt-recovery.json` file properly. If this file has any errors, recovery will be impossible in any circumstance.

IMHO, if we can understand the ways in which `aqt` fails well enough to print an `aqt-recovery.json` file properly on every failure, we should be able to prevent `aqt` from failing in the first place, and this feature would be mostly unnecessary. Maybe it would be useful if you have a really bad network connection or something.

Do any of the popular package managers (apt, dnf, pacman, brew, etc) have features like this that we could borrow ideas from? I don't remember seeing anything like this. I think it would take a lot of engineering effort to develop something like this, and I wonder if that effort would be better spent working on other features.

**Comment by KOLANICH** [comment]
As usual: a journal, rigorously keeping track all the actions to be done and that are done.
